### PR TITLE
Generalize "google_classroom_takeover" logic to work for any oauth section

### PR DESCRIPTION
**Pull request by Hamms**
This logic was originally written to support the fact that google classroom accounts can be created without an identifying email, under the belief that that was true only of google classroom imports. Turns out, that's increasingly true for other logic providers so we want this to be a little more generalizable.

Also rewrite `allows_silent_takeover` for readability.

### CODE REVIEW

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -446,10 +447,12 @@ def sign_in_user(user)
   end
 
   def allows_silent_takeover(oauth_user, auth_hash)
-    allow_takeover = auth_hash.provider.present?
-    allow_takeover &= AuthenticationOption::SILENT_TAKEOVER_CREDENTIAL_TYPES.include?(auth_hash.provider.to_s)
+    return false unless auth_hash.provider.present?
+    return false unless AuthenticationOption::SILENT_TAKEOVER_CREDENTIAL_TYPES.include?(auth_hash.provider.to_s)
+    return false if oauth_user.persisted?
+
     lookup_user = User.find_by_email_or_hashed_email(oauth_user.email)
```
Comment (maddiedierker) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb:454]: i think we should add a lookup for users/authentication_options with the same uid/provider here

Comment (maddiedierker) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb:454]: and on [line 363](https://github.com/code-dot-org/code-dot-org/blob/4e61395f8b758cc1efb6188d393d156e7690a7cf/dashboard/app/controllers/omniauth_callbacks_controller.rb#L363)

**Comment by maddiedierker** [comment]
adding @islemaster to get another pair of eyes on this since elijah and i thought through this change together. brad, would you mind taking a look at this to make sure the approach is reasonable?

File: dashboard/app/controllers/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (Hamms) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb:454]: I'm actually rethinking that; that's what we settled on yesterday, but I'm digging into it and I'm starting to think that takeovers should _only_ be for the situation where a user already exists with the given email, and that we should resolve the situation where a user already exists with the matching oauth information by doing a better job of preventing the second account from being created in the first place.

Comment (maddiedierker) [reply dashboard/app/controllers/omniauth_callbacks_controller.rb:454]: ah, okay, i think that sounds totally reasonable

**Review by islemaster — APPROVED** [review]
Yes! I love this approach, and fewer special cases - I remember suspecting we'd generalize this eventually back when I was in this code more.

**Pull request merged**